sdks/go/pkg/beam/runners/universal/runnerlib/stage.go

```diff
@@ -17,27 +17,127 @@ package runnerlib
 	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+	"google.golang.org/grpc"
 )
 
 // Stage stages the worker binary and any additional files to the given
 // artifact staging endpoint. It returns the retrieval token if successful.
-func Stage(ctx context.Context, id, endpoint, binary, st string, files ...artifact.KeyedFile) (retrievalToken string, err error) {
+func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken string, err error) {
 	ctx = grpcx.WriteWorkerID(ctx, id)
 	cc, err := grpcx.Dial(ctx, endpoint, 2*time.Minute)
 	if err != nil {
 		return "", errors.WithContext(err, "connecting to artifact service")
 	}
 	defer cc.Close()
 
+	err = StageViaPortableApi(ctx, cc, binary, st)
+
+	if err == nil {
+		return "", err
+	} else {
+		return StageViaLegacyApi(ctx, cc, binary, st)
+	}
+}
+
+func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) error {
+	client := jobpb.NewArtifactStagingServiceClient(cc)
+
+	stream, err := client.ReverseArtifactRetrievalService(context.Background())
+	if err != nil {
+		return err
+	}
+
+	if err := stream.Send(&jobpb.ArtifactResponseWrapper{StagingToken: st}); err != nil {
+		return err
+	}
+
+	for {
+		in, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		switch request := in.Request.(type) {
+		case *jobpb.ArtifactRequestWrapper_ResolveArtifact:
+			err = stream.Send(&jobpb.ArtifactResponseWrapper{
+				Response: &jobpb.ArtifactResponseWrapper_ResolveArtifactResponse{
+					&jobpb.ResolveArtifactsResponse{
+						Replacements: request.ResolveArtifact.Artifacts,
+					},
+				}})
+			if err != nil {
+				return err
+			}
+
+		case *jobpb.ArtifactRequestWrapper_GetArtifact:
+			switch typeUrn := request.GetArtifact.Artifact.TypeUrn; typeUrn {
+			case graphx.URNArtifactGoWorker:
+				StageFile(binary, stream)
+
+			default:
+				return errors.Errorf("request has unexpected artifact type %s", typeUrn)
+			}
+
+		default:
+			return errors.Errorf("request has unexpected type %T", request)
+		}
+	}
+}
+
+func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
+	fd, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer fd.Close()
+
+	data := make([]byte, 1<<20)
+	for {
+		n, err := fd.Read(data)
+		if n > 0 {
+			sendErr := stream.Send(&jobpb.ArtifactResponseWrapper{
+				Response: &jobpb.ArtifactResponseWrapper_GetArtifactResponse{
+					&jobpb.GetArtifactResponse{
+						Data: data[:n],
+					},
+				}})
+
+			if sendErr != nil {
+				return errors.Wrap(sendErr, "chunk send failed")
+			}
+		}
+
+		if err == io.EOF {
+			sendErr := stream.Send(&jobpb.ArtifactResponseWrapper{
+				IsLast: true,
+				Response: &jobpb.ArtifactResponseWrapper_GetArtifactResponse{
+					&jobpb.GetArtifactResponse{},
+				}})
+			return sendErr
+		}
+
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
I'm surprise this compiles TBH. Does the compiler detect there are no non-return escapes from the for loop?
I would have expected `return nil` to be required.

Edit: Apparently it does. https://play.golang.org/p/n0SGl6WzJYa Nifty!